---
title: "scDown: A pipeline to carry out the scRNASeq downstream analysis"
author: |
    | 
date: "`r format(Sys.Date(), '%m/%d/%Y')`"
output: 
  rmarkdown::html_document:
    highlight: pygments
    toc: true
    toc_float: true 
    #number_sections: true  ## if you want number sections at each table header
    #theme: united  # many options for theme, this one is my favorite.
#bibliography: library.bib
fig_width: 8 
fig_height: 5 
vignette: >
  %\VignetteIndexEntry{scDown_moncole}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  echo = TRUE,
  message = FALSE,
  warning = FALSE,
  fig.width=8, fig.height=5
)

```


## Introduction

Single-cell transcriptome sequencing (sc-RNA-seq) experiments allow us to discover new cell types and help us understand how they arise in development. The Monocle 3 package provides a toolkit for analyzing single-cell gene expression experiments. In our workflow, we provide a workflow with a single function that carries out the pre-processing, trajectory analysis, and downstream differential expression analysis. In addition to that, we also generates various high quality plots that are very crucial for the trajectory analysis.


## Default moncole3 workflow with the scRNAseq data

Users can load their scRNASeq data which they would like to run the trajectory analysis. To run the trajectory analysis, the users have to provide the scRNASeq data and species. By default, we would run the trajectory analysis on the whole dataset, identify the root nodes using potency method, and run trajectory analysis using moncole3.

```{r, message = FALSE,eval=FALSE}
library(scDown)

# Seurat object
fpath <- system.file("extdata", "p238_testdata.rds", package="scDown")
scRNAseq_seuratobj <- readRDS(fpath)
# Define the species: either "mouse" or "human"
species <- "human"
run_monocle3(seurat=scRNAseq_seuratobj,species=species)

```


## Comparison of trajectory between different conditions

```{r, message = FALSE,eval=FALSE}
library(scDown)

nDim <- 30        # dimension to use for PCA, if object has been clustered by Seurat previously, this dimension can be the same with dimension used in Seurat clustering workflow.
conditions_all <- c("CON","FXPM","FXS")
# a string specifying a column in seurat_object@meta.data that contains the condition names, or NULL when no condition is filled in
# EX). metadata_column <- "orig.ident"
colData_name <- "orig.ident"
run_monocle3(seurat=scRNAseq_seuratobj,species=species,nDim = nDim,conditions = conditions_all,condition_metadata=colData_name)

```

## Comparison of trajectories between different cell types groups

```{r, message = FALSE,eval=FALSE}
library(scDown)

nDim <- 30        # dimension to use for PCA, if object has been clustered by Seurat previously, this dimension can be the same with dimension used in Seurat clustering workflow.
cell_types_monocle<-list(c("OPC","MOL"),c("OL I","OL II"))
run_monocle3(seurat=scRNAseq_seuratobj,species=species,nDim = nDim,conditions = conditions_all,condition_metadata=colData_name, celltype_groups = cell_types_monocle)

```

## Regression analysis between various groups in the scRNASeq data

```{r, message = FALSE,eval=FALSE}
library(scDown)

DEG_distribution <- "quasipoisson"
DEG_models <- c("orig.ident")
run_monocle3(seurat=scRNAseq_seuratobj,species=species,nDim = nDim,conditions = conditions_all, condition_metadata=colData_name, deg_method=DEG_distribution, metadata_deg_model=DEG_models)

```

## Output folders and files explained
All outputs of the pipeline will be in the **results** folder, which is divided into:

* csv
* images
* figures
* rds

The **csv** subfolder contains:

- **monocle**: full DEG analysis results and from them the significant DEGs per specified model and trajectory.
  
The **images** subfolder contains :

- **cellDistribution**: cell and cell type distribution density plots plus histogram per (subsetted) object and condition.

- **DEG**: 
    
    * significant DEGs found by regression analysis on predefined models (defined in universal variable @DEG_models), visualized in feature plot and violin plot. _This is a general linear regression method, and might work better for models that have continuous values_.
    * significant DEGs found by graph auto-correlation analysis along the trajectory/pseudotime, visualized in feature plot, and plotted their expression level changes along pseudotime.
        * These trajectory-variable genes are subsetted to run DEG analysis with regression again, in an effort to identify genes that are differentially expressed along pseudotime AND between two conditions.

    * **pseudotime**: umap by cell types, partitions, learned trajectory, and pseudotime.

The **rds** subfolder contains:

- **rds**: Rds files with the monocle object with the trajectory results 


